src/k8s-engine/client.js

```diff
@@ -48,173 +47,321 @@ class ErrorSuppressingStdin extends events.EventEmitter {
+  }
+
+  /**
+   * Check if a fowarding entry already exists.
+   * @param {string} namespace The namespace to forward to.
+   * @param {string} endpoint The endpoint in the namespace to forward to.
+   * @param {number} port The port to forward to on the endpoint.
+   */
+  has(namespace, endpoint, port) {
+    return super.has(`${namespace}/${endpoint}:${port}`);
+  }
+
+  /**
+   * Iterate through the entries.
+   * @returns {IterableIterator<[string, string, number, net.Server]>}
+   */
+  *[Symbol.iterator]() {
+    const iter = super[Symbol.iterator]();
+    for (const [key, server] of iter) {
+      const [namespace, endpoint, port] = /^([^/]+)\/([^:]+):(\d+)$/.exec(key);
+      yield [namespace, endpoint, parseInt(port), server];
+    }
+  }
+}
+
 /**
  * KubeClient is a Kubernetes client that will _only_ manage the cluster we spin
  * up internally.  The user should call initialize() once the cluster has been
  * created.
  */
-class KubeClient {
-    #kubeconfig = new k8s.KubeConfig();
-    /**
-     * @type k8s.PortForward?
-     */
-    #forwarder = null;
-    /**
-     * @type net.Server?
-     */
-    #server = null;
-    #shutdown = false;
+class KubeClient extends events.EventEmitter {
+  #kubeconfig = new k8s.KubeConfig();
+  /**
+   * @type k8s.PortForward?
+   */
+  #forwarder = null;
 
-    /**
-     * initialize the KubeClient so that we are ready to talk to it.
-     */
-    constructor() {
-      this.#kubeconfig.loadFromDefault();
-      this.#kubeconfig.currentContext = 'rancher-desktop';
-      this.#forwarder = new k8s.PortForward(this.#kubeconfig, true);
-      this.#shutdown = false;
-    }
+  #shutdown = false;
 
-    // Notify that the client the underlying Kubernetes cluster is about to go
-    // away, and we should remove any pending work.
-    destroy() {
-      this.#shutdown = true;
-      this.#server?.close();
-      this.#server = null;
-    }
+  /**
+   * Kubernetes services across all namespaces.
+   * @type {k8s.ListWatch<k8s.V1Service>}
+   */
+  #services;
 
-    /**
-     * @type k8s.CoreV1Api
-     */
-    get #coreV1API() {
-      this.#_coreV1API ||= this.#kubeconfig.makeApiClient(k8s.CoreV1Api);
-      return this.#_coreV1API;
+  /**
+   * Active port forwarding servers.  This records the desired state: if an
+   * entry exists, then we want to set up port forwarding for it.
+   */
+  #servers = new ForwardingMap();
+
+  /**
+   * initialize the KubeClient so that we are ready to talk to it.
+   */
+  constructor() {
+    super();
+    this.#kubeconfig.loadFromDefault();
+    this.#kubeconfig.currentContext = 'rancher-desktop';
+    this.#forwarder = new k8s.PortForward(this.#kubeconfig, true);
+    this.#shutdown = false;
+
+    // Set up a watch for services
+    // Since the watch API we have _doesn't_ notify us when things have
+    // changed, we'll need to do some trickery and wrap the underlying watcher
+    // with our own code.
+    const watch = new k8s.Watch(this.#kubeconfig);
```

Comment:
Hmm, that quite right, since we're not watching for changes to the client, we're watching for changes to the cluster through the client. How do you feel about `k8sWatch`? Or `watcher`?

Reply:
`k8sWatch` works for me